### PR TITLE
improvements to headroom calculation progress tracking

### PR DESCRIPTION
**Pull request by dmy3k**
This PR will allow tracking progress of headroom calculation programatically.
Here is usage example

```python
from gridcapacity.capacity_analysis import CapacityAnalyser
from gridcapacity.config import ConfigModel
from gridcapacity.output import json_dump_kwargs

capacity_analyser = CapacityAnalyser(**kw)  # kw provided outside example
headroom = []
generate, total = capacity_analyser.create_buses_headroom_generator()

for i, (bus_headroom, powerflow_count) in enumerate(generate()):
    print(
        f"progress={i}/{total}, powerflow_count={powerflow_count}"
    )
    headroom.append(bus_headroom)
```

### CODE REVIEW

File: gridcapacity/capacity_analysis.py [hunk 1]
```diff
@@ -210,24 +209,46 @@ def handle_empty_contingency_scenario(
 
     def buses_headroom(self) -> Headroom:
         """Return actual load and max additional PQ power in MVA for each bus"""
-        buses: Buses = Buses()
         console.print("Analysing headroom", style="blue")
+
+        generate, total = self.create_buses_headroom_generator()
+        headroom: List[BusHeadroom] = []
```
Comment (dvlazariev) [inline gridcapacity/capacity_analysis.py:215]:
```suggestion
        headroom: list[BusHeadroom] = []
```

File: gridcapacity/capacity_analysis.py [hunk 2]
```diff
@@ -19,9 +19,7 @@
 from collections import OrderedDict, defaultdict
 from collections.abc import Collection
 from dataclasses import dataclass
-from typing import Final, Iterator, Optional
-
-from tqdm import tqdm
+from typing import Final, Generator, Iterator, List, Optional, Tuple
```
Comment (dvlazariev) [inline gridcapacity/capacity_analysis.py:22]:
```suggestion
from typing import Final, Generator, Iterator, Optional, Tuple
```

File: gridcapacity/capacity_analysis.py [hunk 3]
```diff
@@ -210,24 +209,46 @@ def handle_empty_contingency_scenario(
...
+        buses: Buses = Buses()
+        total = (
+            len(buses)
+            if self._selected_buses_ids is None
+            else len(self._selected_buses_ids)
+        )
+        PowerFlows.reset_count()
+        ViolationsStats.reset()
+
+        def generate():
```
Comment (dvlazariev) [inline gridcapacity/capacity_analysis.py:241]:
```suggestion
        def generate() -> Generator[tuple[BusHeadroom, int]]:
```

File: gridcapacity/capacity_analysis.py [hunk 4]
```diff
@@ -210,24 +209,46 @@ def handle_empty_contingency_scenario(
...
-            )
+            for bus_headroom, _ in generate():
+                progress.postfix[0]["bus_number"] = bus_headroom.bus.number
+                progress.postfix[0]["power_flows"] = PowerFlows.count
+                progress.update()
+                headroom.append(bus_headroom)
+
+        return tuple(headroom)
+
+    def create_buses_headroom_generator(self):
```
Comment (dvlazariev) [inline gridcapacity/capacity_analysis.py:231]:
```suggestion
    def create_buses_headroom_generator(self) -> Callable[Generator[tuple[BusHeadroom, int]]]:
```
I'm not sure about this one, please check.

File: gridcapacity/capacity_analysis.py [hunk 5]
```diff
@@ -210,24 +209,46 @@ def handle_empty_contingency_scenario(
...
-                if self._selected_buses_ids is None
-                or bus.number in self._selected_buses_ids
-            )
+            for bus_headroom, _ in generate():
+                progress.postfix[0]["bus_number"] = bus_headroom.bus.number
+                progress.postfix[0]["power_flows"] = PowerFlows.count
+                progress.update()
+                headroom.append(bus_headroom)
+
+        return tuple(headroom)
```
Comment (dvlazariev) [inline gridcapacity/capacity_analysis.py:229]: Please, consider changing the `Headroom` from `tuple` to a `list` and don't cast here.
```suggestion
        return headroom
```

File: gridcapacity/capacity_analysis.py (hunk 4 above)
Comment (dmy3k) [reply gridcapacity/capacity_analysis.py:231]: Suggested type is incorrect. 
The right one would be `tuple[() -> Generator[tuple[BusHeadroom, Any], Any, None], int]`
However it is correctly inferred from return value. I think there is no need to specify it explicitly.

File: gridcapacity/capacity_analysis.py (hunk 5 above)
Comment (dmy3k) [reply gridcapacity/capacity_analysis.py:229]: Type casting is made to be compatible with types set in the call chain of current method.
Expected type is `Headroom` which is tuple.

Comment (dmy3k) [reply gridcapacity/capacity_analysis.py]: Done.

**Pull request merged**